### PR TITLE
NAICS endpoint and test

### PR DESCRIPTION
**Pull request by russellbodine**
https://federal-spending-transparency.atlassian.net/browse/DS-1340

Gist reviewed by ( @ejgullo ):
- [x] https://gist.github.com/russellbodine/3c7cd4652db732331ec006bcfbb2c30d

### CODE REVIEW

**Comment by russellbodine** [comment]
https://gist.github.com/russellbodine/3c7cd4652db732331ec006bcfbb2c30d

**Comment by russellbodine** [comment]
DS 1340 
https://federal-spending-transparency.atlassian.net/browse/DS-1340

File: usaspending_api/references/urls_autocomplete.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+    NAICSAutocompleteViewSet, RecipientAutocompleteViewSet
 
 
 urlpatterns = [
     url(r'^awarding_agency', AwardingAgencyAutocompleteViewSet.as_view()),
     url(r'^budget_function', BudgetFunctionAutocompleteViewSet.as_view()),
     url(r'^funding_agency', FundingAgencyAutocompleteViewSet.as_view()),
-    url(r'^recipient/', RecipientAutocompleteViewSet.as_view())
+    url(r'^naics', NAICSAutocompleteViewSet.as_view()),
+    url(r'^recipient/', RecipientAutocompleteViewSet.as_view()),
```
Comment (nmonga91) [inline usaspending_api/references/urls_autocomplete.py]: does the extra comma here mess with anything? probably should remove it anyways to denote the end of the list

File: usaspending_api/references/views_v2/autocomplete.py [hunk 2]
```diff
@@ -106,6 +105,34 @@ def post(self, request):
...
 
+class NAICSAutocompleteViewSet(BaseAutocompleteViewSet):
+
+    def post(self, request):
+        """Return all budget function/subfunction titles matching the provided search text"""
+
+        search_text, limit = self.get_request_payload(request)
+
+        # get relevant TransactionContracts
+        queryset = TransactionContract.objects.filter(naics__isnull=False, naics_description__isnull=False)
```
Comment (nmonga91) [inline usaspending_api/references/views_v2/autocomplete.py]: Is it possible for one of the fields to be populated while the other is not? Want to make sure we're not restricting our search more than necessary.

File: usaspending_api/references/views_v2/autocomplete.py [hunk 3]
```diff
@@ -106,6 +105,34 @@ def post(self, request):
...
+            TrigramSimilarity('naics_description', search_text)))\
+            .distinct().order_by('-similarity')
+
+        naics_exact_match_queryset = queryset.filter(similarity=1.0)
+        if naics_exact_match_queryset.count() > 0:
+            queryset = naics_exact_match_queryset
+
+        results_set = list(queryset.values('naics', 'naics_description')[:limit])
+        response['results'] = results_set
+        response['counts'] = len(results_set)
```
Comment (nmonga91) [inline usaspending_api/references/views_v2/autocomplete.py]: Is the count still necessary for this response? I know most of the v1 endpoints have them, but if it's not needed, we should remove it

File: usaspending_api/references/views_v2/autocomplete.py [hunk 4]
```diff
@@ -106,6 +105,34 @@ def post(self, request):
...
+        queryset = queryset.annotate(similarity=Greatest(
+            TrigramSimilarity('naics', search_text),
+            TrigramSimilarity('naics_description', search_text)))\
+            .distinct().order_by('-similarity')
+
+        naics_exact_match_queryset = queryset.filter(similarity=1.0)
+        if naics_exact_match_queryset.count() > 0:
+            queryset = naics_exact_match_queryset
+
+        results_set = list(queryset.values('naics', 'naics_description')[:limit])
```
Comment (nmonga91) [inline usaspending_api/references/views_v2/autocomplete.py]: If we're not longer relying on custom formatting or adding counts in the response, we could still leverage serializers here. Depends on the decision made about the counts in the response. The recipient autocomplete view set follows that example.

**Comment by ejgullo** [comment]
Signing off on gist.

File: usaspending_api/references/urls_autocomplete.py (hunk 1 above)
Comment (russellbodine) [inline usaspending_api/references/urls_autocomplete.py]: it doesnt mess with anything, but im removing it!

File: usaspending_api/references/views_v2/autocomplete.py (hunk 2 above)
Comment (russellbodine) [inline usaspending_api/references/views_v2/autocomplete.py]: regardless, trigram search will break without both i think?

File: usaspending_api/references/views_v2/autocomplete.py (hunk 4 above)
Comment (russellbodine) [inline usaspending_api/references/views_v2/autocomplete.py]: serializers tend to add headache for no gain. Im gonna stick to custom responses.

File: usaspending_api/references/views_v2/autocomplete.py (hunk 3 above)
Comment (russellbodine) [inline usaspending_api/references/views_v2/autocomplete.py]: removing it.

File: usaspending_api/references/views_v2/autocomplete.py (hunk 2 above)
Comment (russellbodine) [inline usaspending_api/references/views_v2/autocomplete.py]: its not possible

**Pull request merged**